cli/src/generate/templates/setup.py

```diff
@@ -0,0 +1,37 @@
+from setuptools import setup, find_packages
+from cffi import FFI
+
+ffibuilder = FFI()
+
+ffibuilder.set_source(
+    "tree_sitter_PARSER_NAME._tree_sitter_PARSER_NAME",
+    """
+    #include "../../../src/tree_sitter/parser.h"
+    extern const TSLanguage *tree_sitter_PARSER_NAME();
+    """,
+    # if you have a scanner, you need to include it here too
+    sources=["../../src/parser.c"],
+    include_dirs=[
+        "../../src"
+    ],
+)
+
+ffibuilder.cdef("""
+    typedef struct TSLanguage TSLanguage;
+    const TSLanguage *tree_sitter_PARSER_NAME(void);
+""")
+
+if __name__ == "__main__":
+    ffibuilder.compile(verbose=True)
+    setup(
+        name='tree-sitter-PARSER_NAME',
+        version='0.0.1',
+        description='CAMEL_PARSER_NAME grammar for tree-sitter',
+        url='http://github.com/tree-sitter/tree-sitter-PARSER_NAME',
+        license='MIT',
+        packages=find_packages(),
+        package_data={
+            'tree_sitter_PARSER_NAME': ['_tree_sitter_PARSER_NAME.cpython-311-darwin.so'],
```

Comment:
this name feels very bad, since you assume a lot of things (OS, python version).

Reply:
You can get this suffix for the current Python interpreter via `distutils`
```
import distutils.sysconfig
distutils.sysconfig.get_config_var('EXT_SUFFIX')
```

Reply:
You should be very careful of using `distutils`, since it was removed in python 3.12. There is a continuation as part of `setuptools`, so I recommend to verify it carefully that it works on python 3.12.

Reply:
As package data, this `.so` shouldn't require any suffix as it is not thought to be importable as Python module but `dlopen`ed by python-tree-sitter.

Reply:
> this name feels very bad, since you assume a lot of things (OS, python version).

Hm, yeah, how should I go about this to be platform-agnostic?

Reply:
[Documentation](https://cffi.readthedocs.io/en/latest/cdef.html#ffibuilder-compile-etc-compiling-out-of-line-modules)

Looks like you can rename it manually after calling `compile`, or try the setuptools variant. In that same page search for `from setuptools import setup` line and see the example and one example below it. I **think** it should work well.

Reply:
@amaanq I can confirm sadly that the python isn't working when installed into a virtual environment, it can't find the library (even after I changed to match my package_data extension). I suspect it isn't imported into the correct place. Let me try to fix it.

Reply:
Hm, yeah I'm not super familiar with Python's build process, running `python setup.py install --user` gives me this file structure:

```
$ fd -t f
build/lib/tree_sitter_capnp/__init__.py
dist/tree_sitter_capnp-0.0.1-py3.11.egg
setup.py
tree_sitter_capnp/__init__.py
tree_sitter_capnp/_tree_sitter_capnp.c
tree_sitter_capnp/_tree_sitter_capnp.cpython-311-x86_64-linux-gnu.so
tree_sitter_capnp/_tree_sitter_capnp.o
tree_sitter_capnp.egg-info/PKG-INFO
tree_sitter_capnp.egg-info/SOURCES.txt
tree_sitter_capnp.egg-info/dependency_links.txt
tree_sitter_capnp.egg-info/not-zip-safe
tree_sitter_capnp.egg-info/top_level.txt
```

Reply:
And then, once you install it in your venv, you can import a language:

```sh
Python 3.11.5 (main, Aug 28 2023, 20:02:58) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tree_sitter_capnp
>>> tree_sitter_capnp.language()
<cdata 'struct TSLanguage *' 0x7f741c5f5d00>
>>> 
```

Which is just the raw TSLanguage struct which is meant to be used in py-tree-sitter

Reply:
It seems like ffibuilder.compile returns the file name so...
```py
if __name__ == "__main__":
    file = ffibuilder.compile(verbose=True)
    setup(
        name='tree-sitter-capnp',
        version='0.0.1',
        description='Capnp grammar for tree-sitter',
        url='http://github.com/tree-sitter/tree-sitter-capnp',
        license='MIT',
        packages=find_packages(),
        package_data={
            'tree_sitter_capnp': [file.split('/')[-1]]
        },
        zip_safe=False
    )
```

This might work dynamically

Reply:
So, yeah, a little bit of background for python, since a lot has changed since, because of PEP517.

1. Don't call directly `setup.py` with commands - this is deprecated
2. Use `build` package to build a wheel package from the directory. For example: `python3.12 -m build -w` (which by the way fails because of missing `setup_requires=["cffi>=1.0.0"],`
3. Install it into a venv to verify it working:

```
python3.12 -m venv /tmp/venv
source /tmp/venv/bin/activate
pip install dist/*.whl
python
```

Reply:
I see - I get the same error. It seems setup_requires is deprecated in favor of using a pyproject.toml to specify requirements, should this be the way to go? When I do that and specify cffi and setuptools as requirements, installation works just fine

Reply:
@amaanq Yes, you should use pyproject.toml file.

Also, I'm sorry, but I don't think I'll have time for OSS work in the near half-week, so if my input is a must, but you want to merge it, feel free to continue and I can return after to python bindings.

Reply:
No worries! Your insight here was already very helpful for properly building a python package lately.

One last question if at all possible - The build commands builds a .whl file, is that fine? I do have to run pip install dist/tree_sitter_lang-ver-py3-none-any.whl to actually have the package available, I'm guessing that's normal

Reply:
By some recommendations, I'll investigate the possibility of using meson-python for this. It should bring us the best interface and result.

> One last question if at all possible - The build commands builds a .whl file, is that fine? I do have to run pip install dist/tree_sitter_lang-ver-py3-none-any.whl to actually have the package available, I'm guessing that's normal

Yes, this is the "binary artifact" for python packages. "tar.gz" for source and "whl" for binary.

Reply:
My end goal with this is as such, at least for python:

tree-sitter-{LANG} will provide python packages, at least the upstream grammars, and with that will distribute the packages to PyPI, where a user can then run `pip install tree-sitter-c` to instantly fetch the compiled? (or should they compile it on install?) parser. The language() function will basically do what `Language()` does in `py-tree-sitter`